ledger/participant-integration-api/src/test/suite/scala/platform/store/cache/StateCacheSpec.scala

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.cache
+
+import java.util.concurrent.TimeUnit
+
+import com.codahale.metrics.MetricRegistry
+import com.daml.caching.{CaffeineCache, ConcurrentCache}
+import com.daml.logging.LoggingContext
+import com.daml.metrics.Metrics
+import com.github.benmanes.caffeine.cache.Caffeine
+import org.mockito.MockitoSugar
+import org.scalatest.Assertion
+import org.scalatest.concurrent.Eventually
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import scala.concurrent.duration.{FiniteDuration, _}
+import scala.concurrent.{Await, ExecutionContext, Future, Promise}
+import scala.util.Success
+
+class StateCacheSpec extends AnyFlatSpec with Matchers with MockitoSugar with Eventually {
+  private implicit val loggingContext: LoggingContext = LoggingContext.ForTesting
+  private val metrics = new Metrics(new MetricRegistry)
+
+  behavior of s"${classOf[StateCache[_, _]].getSimpleName}.putAsync"
+
+  it should "asynchronously store the update" in {
+    implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.global
+
+    val cache = mock[ConcurrentCache[String, String]]
+    val stateCache = StateCache[String, String](cache, metrics)
+
+    val asyncUpdatePromise = Promise[String]
+    stateCache.putAsync("key", 1L, asyncUpdatePromise.future)
+    asyncUpdatePromise.completeWith(Future.successful("value"))
+
+    verify(cache).put("key", "value")
+    // Async update should not insert in the cache
+    verifyNoMoreInteractions(cache)
+  }
+
+  it should "store the latest key update in face of conflicting pending updates" in {
+    implicit val ec: ExecutionContext = scala.concurrent.ExecutionContext.global
+    val `number of competing updates` = 1000L
+    val `number of keys in cache` = 100L
+
+    val stateCache = buildStateCache(`number of keys in cache`)
+
+    val insertions = prepare(`number of competing updates`, `number of keys in cache`)
+
+    val (insertionFutures, insertionDuration) = insertTimed(stateCache)(insertions)
+
+    insertionDuration should be < 1.second
+
+    insertions.foreach { case (_, (promise, value, _)) =>
+      promise.complete(Success(value))
+    }
+
+    Await.result(Future.sequence(insertionFutures.toVector), 5.second).nonEmpty shouldBe true
+    assertCacheElements(stateCache)(insertions, `number of competing updates`)
+  }
+
+  it should "putAsync 100_000 values for the same key in 1 second" in {
```

Comment:
Could this test become flaky in constrained CI elements?

Reply:
Yes it can but I don't think more than other timeout-sensitive tests. I've set the limit at 100K quite conservatively (on my machine I was getting 300-400K in 1 second).